### PR TITLE
[CBRD-24478]  Make sure that the index is not skewed by duplicate keys.(After Change Spec.)

### PR DESCRIPTION
**Pull request by ctshim**
http://jira.cubrid.org/browse/CBRD-24478

* Reflect changed requirements
1) FK
    system parameter "deduplicate_fk_level": default value is 32, unchangeable in HIDDEN, SESSION
    FK is automatically specified using the "deduplicate_fk_level" value.
    (To create an FK that does not use this function, you must specify "deduplicate_fk_level=0" and record it in the conf file.)
    FK does not show deduplicate information in output in any case.
   
    Exceptionally, the syntax that can define an FK that does not use this function is the ALTER TABLE ADD CONSTRAINT statement. (Private)
    ex) alter table tbl_s ADD CONSTRAINT FOREIGN KEY (id) WITH DEDUPLICATE=0 References tbl_m(id);
   
         
2) NORMAL INDEX
    system parameter "deduplicate_key_level": Default value is 10, can be changed in SESSION, HA delivery
                     "deduplicate_min_keys": Default value is 2, can be changed in SESSION, HA delivery
    2-1) In case of omitting DEDUPLICATE designation
        Automatically set as "deduplicate_key_level" value
        (If you do not want to apply the deduplicate function by default, set "deduplicate_key_level=0" in cubrid.conf.)
        If "deduplicate_min_keys" is not 0, it is applied only when the number of key fields is less than the set value of "deduplicate_min_keys".
          
    2-2) When DEDUPLICATE is specified
       If level is not specified, 32 is assigned (with deduplicate).
       If level is specified, the specified level is used.
      
    In UnloadDB, SHOW CREATE TABLE, and ;SC commands, if the deduplicate_level of the index is 0, the output is not displayed.

3) Avg_num_value_per_deduplicate_key item added (corresponds to Avg_num_value_per_key)
    * fix avg_rec_len calculation

### CODE REVIEW

File: src/object/class_object.c [hunk 1]
```diff
@@ -8244,6 +8258,30 @@ classobj_check_index_exist (SM_CLASS_CONSTRAINT * constraints, char **out_shared
     case SM_SHARE_INDEX:
       if (out_shared_cons_name != NULL)
 	{
+#if defined(SUPPORT_DEDUPLICATE_KEY_MODE)
+	  if (constraint_type == DB_CONSTRAINT_FOREIGN_KEY)
+	    {
+	      int level;
+	      SM_ATTRIBUTE **attp = existing_con->attributes;
```
Comment (beyondykk9) [inline src/object/class_object.c:8265]: assert (attp != NULL) is not needed?

Comment (ctshim) [reply src/object/class_object.c:8265]: This is checked in the classobj_find_constraint_by_attrs() function performed earlier.
If existing_con exists, it must have an attribute.

**Review by beyondykk9 — APPROVED** [review]

**Review by hyunikn — APPROVED** [review]

File: src/parser/csql_grammar.y [hunk 2]
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
...
+        | '='
+        ;
+
+online_parallel
+	: ONLINE
+	   {{ DBG_TRACE_GRAMMAR(online_parallel, : ONLINE);
+               $$ = 1;  // thread count is 0
+           DBG_PRINT}}
+	| ONLINE PARALLEL opt_equal unsigned_integer
+	   {{ DBG_TRACE_GRAMMAR(opt_ononline_parallelline, |  ONLINE PARALLEL opt_equal unsigned_integer);
```
Comment (youngjinj) [inline src/parser/csql_grammar.y]:
```suggestion
	   {{ DBG_TRACE_GRAMMAR(online_parallel, |  ONLINE PARALLEL opt_equalsign unsigned_integer);
```

File: src/parser/csql_grammar.y [hunk 3]
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
...
+                container_3 ctn;
+		SET_CONTAINER_3(ctn, $3, $1, DEDUPLICATE_MIN_KEYS_UNUSE);
+		$$ = ctn;
+          }
+        ;        
+
+opt_equal
+        : /* empty */
+        | '='
+        ;
```
Comment (youngjinj) [inline src/parser/csql_grammar.y]: Redundant with `opt_equalsign` rule.

File: src/parser/csql_grammar.y [hunk 4]
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
...
+
+online_parallel
+	: ONLINE
+	   {{ DBG_TRACE_GRAMMAR(online_parallel, : ONLINE);
+               $$ = 1;  // thread count is 0
+           DBG_PRINT}}
+	| ONLINE PARALLEL opt_equal unsigned_integer
```
Comment (youngjinj) [inline src/parser/csql_grammar.y]:
```suggestion
	| ONLINE PARALLEL opt_equalsign unsigned_integer
```

File: src/parser/csql_grammar.y [hunk 5]
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
...
+		SET_CONTAINER_3(ctn, 0, $1, DEDUPLICATE_MIN_KEYS_UNUSE);
+		$$ = ctn;
+          }
+        | deduplicate_key_mod_level ',' online_parallel
+          { DBG_TRACE_GRAMMAR(index_with_item_list, | deduplicate_key_mod_level ',' online_parallel ); 
+                container_3 ctn;
+		SET_CONTAINER_3(ctn, $3, $1, DEDUPLICATE_MIN_KEYS_UNUSE);
+		$$ = ctn;
+          }
+        ;
```
Comment (youngjinj) [inline src/parser/csql_grammar.y]: Is there any more elegant way?
```
index_with_item_list  
        : online_parallel
        | online_parallel ',' deduplicate_key_mod_level
        | deduplicate_key_mod_level
        | deduplicate_key_mod_level ',' online_parallel
        ;
```
Even if we add only one new option, it should change like below.
```
index_with_item_list  
        : online_parallel
        | online_parallel ',' deduplicate_key_mod_level
        | online_parallel ',' <new_option>
        | online_parallel ',' deduplicate_key_mod_level ',' <new_option>
        | online_parallel ',' <new_option> ',' deduplicate_key_mod_level
        | deduplicate_key_mod_level
        | deduplicate_key_mod_level ',' online_parallel
        | deduplicate_key_mod_level ',' <new_option>
        | deduplicate_key_mod_level ',' online_parallel ',' <new_option>
        | deduplicate_key_mod_level ',' <new_option> ',' online_parallel
        | <new_option>
        | <new_option> ',' online_parallel
        | <new_option> ',' deduplicate_key_mod_level
        | <new_option> ',' online_parallel ',' deduplicate_key_mod_level
        | <new_option> ',' deduplicate_key_mod_level ',' online_parallel
        ;
```
The following code is used in postgresql.
```
alterfunc_opt_list:
                        /* At least one option must be specified */
                        common_func_opt_item
                        | alterfunc_opt_list common_func_opt_item 
                ;

common_func_opt_item:
                        CALLED ON NULL_P INPUT_P
                        | RETURNS NULL_P ON NULL_P INPUT_P
                        | STRICT_P
                        | IMMUTABLE
                        | STABLE
                        | VOLATILE
                        | EXTERNAL SECURITY DEFINER
                        | EXTERNAL SECURITY INVOKER
                        | SECURITY DEFINER
                        | SECURITY INVOKER
                        | LEAKPROOF
                        | NOT LEAKPROOF
                        | COST NumericOnly
                        | ROWS NumericOnly
                        | SUPPORT any_name
                        | FunctionSetResetClause
                        | PARALLEL ColId
                ;

static bool
compute_common_attribute(...)
{
        if (strcmp(defel->defname, "volatility") == 0)
        {
                if (is_procedure)
                        goto procedure_error;
                if (*volatility_item)
                        goto duplicate_error;

                *volatility_item = defel;
        }
        ...
duplicate_error:
        ereport(ERROR,
                        (errcode(ERRCODE_SYNTAX_ERROR),
                         errmsg("conflicting or redundant options"),
                         parser_errposition(pstate, defel->location)));
        return false;
...
```

File: src/parser/csql_grammar.y [hunk 6]
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
...
+opt_fk_deduplicate_off
+        : /* empty */
+          { DBG_TRACE_GRAMMAR(opt_fk_deduplicate_off, : );
+            $$ = false; }
+        | WITH DEDUPLICATE_ '=' unsigned_integer
+          { DBG_TRACE_GRAMMAR(opt_fk_deduplicate_off,  | DEDUPLICATE_ '=' unsigned_integer); 
+           /* It can be entered through the table constraint clause of "CREATE TABLE" and the add constraint clause of "ALTER TABLE".
+            * If it is entered in the "CREATE TABLE" statement, an error is processed.
+           */          
+            if(parser_is_null_alter_node())
```
Comment (youngjinj) [inline src/parser/csql_grammar.y:21569]: Is the purpose of preventing developers from adding this rule elsewhere?

File: src/parser/csql_grammar.y [hunk 7]
```diff
@@ -21587,45 +21557,129 @@ opt_encrypt_algorithm
...
+        ;
+
+opt_index_with_clause
+        : /* empty */
+          { DBG_TRACE_GRAMMAR(opt_index_with_clause, : );
+            container_3 ctn;
+            memset(&ctn, 0x00, sizeof(ctn));
+            ctn.c1 = 0; // not online
+            ctn.c2 = prm_get_integer_value (PRM_ID_DEDUPLICATE_KEY_LEVEL);
+            ctn.c3 = prm_get_integer_value (PRM_ID_DEDUPLICATE_MIN_KEYS);
```
Comment (youngjinj) [inline src/parser/csql_grammar.y]: The result below seems to be the unintended result.
```
$ csql -u dba demodb -S

csql> ;edit
drop table if exists t1;
create table t1 (c1 int, c2 int, c3 int, c4 int, c5 int);
set system parameters 'deduplicate_key_level=11';
create index i1_level11 on t1 (c1);
set system parameters 'deduplicate_key_level=12';
create index i2_level12 on t1 (c2);
set system parameters 'deduplicate_key_level=13';
create index i3_level13 on t1 (c3);
set system parameters 'deduplicate_key_level=14';
create index i4_level14 on t1 (c4);
set system parameters 'deduplicate_key_level=15';
create index i5_level15 on t1 (c5);

csql> ;x
csql> ;sc t1
```
Expected Result
```
 <Constraints>

     INDEX i1_level11 ON dba.t1 (c1) WITH DEDUPLICATE=11
     INDEX i2_level12 ON dba.t1 (c2) WITH DEDUPLICATE=12
     INDEX i3_level13 ON dba.t1 (c3) WITH DEDUPLICATE=13
     INDEX i4_level14 ON dba.t1 (c4) WITH DEDUPLICATE=14
     INDEX i5_level15 ON dba.t1 (c5) WITH DEDUPLICATE=15
```
Actual Result
```
 <Constraints>

     INDEX i1_level11 ON dba.t1 (c1) WITH DEDUPLICATE=10
     INDEX i2_level12 ON dba.t1 (c2) WITH DEDUPLICATE=10
     INDEX i3_level13 ON dba.t1 (c3) WITH DEDUPLICATE=10
     INDEX i4_level14 ON dba.t1 (c4) WITH DEDUPLICATE=10
     INDEX i5_level15 ON dba.t1 (c5) WITH DEDUPLICATE=10
```

File: src/parser/csql_grammar.y (hunk 6 above)
Comment (ctshim) [reply src/parser/csql_grammar.y:21569]: For FK, only "with deduplicate=0" can be specified using the ALTER TABLE statement. However, it is not a feature that is open to the general user.For FK, only "with deduplicate=0" can be specified using the ALTER TABLE statement. However, it is not a feature that is open to the general user.

File: src/parser/csql_grammar.y (hunk 2 above)
Comment (ctshim) [reply src/parser/csql_grammar.y]: Thank you

File: src/parser/csql_grammar.y (hunk 3 above)
Comment (ctshim) [reply src/parser/csql_grammar.y]: Thank you, I will fix it.

File: src/parser/csql_grammar.y (hunk 4 above)
Comment (ctshim) [reply src/parser/csql_grammar.y]: OK,

File: src/parser/csql_grammar.y (hunk 7 above)
Comment (ctshim) [reply src/parser/csql_grammar.y]: Good check.
I will fix it.  thanks.

File: src/parser/csql_grammar.y (hunk 6 above)
Comment (youngjinj) [reply src/parser/csql_grammar.y:21569]: I was mistaken that this rule is not used for create_stmt. Thank you.

File: src/parser/csql_grammar.y (hunk 5 above)
Comment (ctshim) [reply src/parser/csql_grammar.y]: Modifications to be flexible will be carried out as a separate work later.

**Review by youngjinj — APPROVED** [review]

**Review by shparkcubrid — APPROVED** [review]

**Review by hgryoo — APPROVED** [review]

**Pull request merged**